raiden-ts/src/services/epics.ts

```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
+      return resetAllowance$.pipe(
+        mergeMap(() =>
+          tokenContract.functions.approve(
+            userDepositContract.address,
+            bnMax(minimumAllowance, deposit),
+          ),
+        ),
+        assertTx('approve', ErrorCodes.RDN_APPROVE_TRANSACTION_FAILED, { log }),
+      );
+    }),
+    // send setTotalDeposit transaction
+    mergeMap(() => userDepositContract.functions.deposit(address, totalDeposit)),
+    assertTx('deposit', ErrorCodes.RDN_DEPOSIT_TRANSACTION_FAILED, { log }),
+    // retry also txFail errors, since estimateGas can lag behind just-opened channel or
+    // just-approved allowance
+    retryTx(pollingInterval, undefined, undefined, { log }),
+  );
+}
+
+/**
+ * Handles a udcDeposit.request and deposit SVT/RDN to UDC
+ *
+ * @param action$ - Observable of udcDeposit.request actions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - Epics dependencies
+ * @param deps.userDepositContract - UserDeposit contract instance
+ * @param deps.getTokenContract - Factory for Token contracts
+ * @param deps.address - Our address
+ * @param deps.log - Logger instance
+ * @param deps.signer - Signer
+ * @param deps.main - Main Signer/Address
+ * @param deps.config$ - Config observable
+ * @returns - Observable of udcDeposit.failure|udcDeposit.success actions
+ */
+export const udcDepositEpic = (
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { userDepositContract, getTokenContract, address, log, signer, main, config$ }: RaidenEpicDeps,
+): Observable<udcDeposit.failure | udcDeposit.success> => {
+  const svtToken = userDepositContract.functions.token() as Promise<Address>;
+  return action$.pipe(
+    filter(udcDeposit.request.is),
+    concatMap((action) =>
+      defer(() => svtToken).pipe(
+        withLatestFrom(config$),
+        mergeMap(([token, config]) => {
+          const { signer: onchainSigner, address: onchainAddress } = chooseOnchainAccount(
+            { signer, address, main },
+            action.payload.subkey ?? config.subkey,
+          );
+          const tokenContract = getContractWithSigner(getTokenContract(token), onchainSigner);
+          const udcContract = getContractWithSigner(userDepositContract, onchainSigner);
+
+          return makeUdcDeposit$(
+            [tokenContract, udcContract],
+            [onchainAddress, address],
+            [action.payload.deposit, action.meta.totalDeposit],
+            config,
+            { log },
+          );
+        }),
+        map(([, receipt]) =>
+          udcDeposit.success(
+            {
+              txHash: receipt.transactionHash,
+              txBlock: receipt.blockNumber,
+              confirmed: undefined, // let confirmationEpic confirm this action
+            },
+            action.meta,
+          ),
+        ),
+        catchError((error) => of(udcDeposit.failure(error, action.meta))),
+      ),
+    ),
```

Comment:
This function is pretty good. It mainly has a single layer of abstraction that could be followed. Most small details are hidden in dedicated functions that are named descriptive enough to understand what they do.